### PR TITLE
replace PFLR DeviceGetContext hardcode with Device::IsRemoteCallAllowed

### PR DESCRIPTION
**Pull request by ppiskorski**
This delegates decision of whether a particular device type
can host a remote call to the device implementation itself.

### CODE REVIEW

**Comment by ppiskorski** [comment]
@googlebot I signed it!
________________________________

File: tensorflow/core/common_runtime/device.cc [hunk 1]
```diff
@@ -54,4 +54,22 @@ DeviceAttributes Device::BuildDeviceAttributes(
   return da;
 }
 
+
+bool Device::IsRemoteCallAllowed() const {
```
Comment (zhangyujing) [inline tensorflow/core/common_runtime/device.cc:58]: Could you avoid the hardcode here?

**Review by zhangyujing — COMMENTED** [review]
Could you please provide some context of this change? Is it a bug fix? The behavior after this change seems changed. Is it expected?

**Comment by qqfish** [comment]
Can you give more context about why you make this change? Does it cause any bugs/issues in your use case?

I think some device types (e.g. composite device) are not supposed to be used by PFLR. Looks like it indicates some bugs in the caller site as well.

Thanks.

**Comment by ppiskorski** [comment]
Hi qqfish
Thanks for comment!
> 
> 
> Can you give more context about why you make this change? Does it cause any bugs/issues in your use case?
> 
> I think some device types (e.g. composite device) are not supposed to be used by PFLR. Looks like it indicates some bugs in the caller site as well.
> 
> Thanks.

This PR is in reference to a [discussion I started on tf.developers](https://groups.google.com/a/tensorflow.org/g/developers/c/q8C3CkMkKew). In short, we are adding support for a hardware accelerator device that has significant speedup when prefetch_to_device is used in the data pipeline. That modifier creates an iterator on the accelerator device that executes the whole pipeline on CPU using RemoteCall which is only allowed for a limted set of devices that's currently hardcoded in ProcessFunctionLibraryRuntime::GetDeviceContext.

My naive proposal with this PR is to extend device interface so each device class can testify if it can be a remote caller or not.

**Comment by qqfish** [comment]
Thanks for the context. It sgtm.

File: tensorflow/core/common_runtime/device.h [hunk 2]
```diff
@@ -178,6 +178,8 @@ class Device : public DeviceBase {
 
   virtual bool IsLocal() const { return true; }
 
+  virtual bool IsRemoteCallAllowed() const = 0;
```
Comment (qqfish) [inline tensorflow/core/common_runtime/device.h:181]: Can you add a comment to this method?

**Comment by gbaned** [comment]
@ppiskorski Can you please check @qqfish's comments and keep us posted ? Thanks!

File: tensorflow/core/common_runtime/device.h (hunk 2 above)
Comment (ppiskorski) [reply tensorflow/core/common_runtime/device.h:182]: Hi qqfish
I've added brief information. Please let me know if you want me to be more elaborate.

File: tensorflow/core/common_runtime/device.cc (hunk 1 above)
Comment (ppiskorski) [reply tensorflow/core/common_runtime/device.cc:58]: Seems straightforward with the exception of TPU_SYSTEM which is configured into VirtualDevice.

But this question led me to discovery of DeviceAttributes. So another approach seems to be to add is_remote_call_supported to DeviceAttribute message and implement IsRemoteCallSupported as a non-virtual attribute getter. Do you think that would be a better solution?

**Comment by gbaned** [comment]
@ppiskorski  Can you please resolve conflicts? Thanks!

**Review by qqfish — APPROVED** [review]

File: tensorflow/compiler/jit/xla_device.h [hunk 3]
```diff
@@ -188,6 +188,8 @@ class XlaDevice : public LocalDevice {
 
   Status RefreshStatus() override TF_LOCKS_EXCLUDED(mu_);
 
+  bool IsRemoteCallAllowed() const override { return true; }
```
Comment (zhangyujing) [inline tensorflow/compiler/jit/xla_device.h:191]: Is this override needed for XlaDevice? also for FakeDevice, XlaCompilationDevice, SingleThreadedCpuDevice?

Comment (ppiskorski) [reply tensorflow/compiler/jit/xla_device.h:191]: Hi zhangyujing Thanks for the comment. I refined the the list of overrides and will push an update shortly.

File: tensorflow/core/grappler/optimizers/function_optimizer.cc [hunk 4]
```diff
@@ -115,7 +115,6 @@ class FakeDevice : public Device {
   FakeDevice(Env* env, const string& device) : Device(env, attr(device)) {}
   explicit FakeDevice(const string& device) : FakeDevice(nullptr, device) {}
   Status Sync() override { return Status::OK(); }
-
```
Comment (zhangyujing) [inline tensorflow/core/grappler/optimizers/function_optimizer.cc:118]: Is this change intended?

File: tensorflow/core/common_runtime/threadpool_device.h [hunk 5]
```diff
@@ -44,6 +44,8 @@ class ThreadPoolDevice : public LocalDevice {
 
   Status Sync() override { return Status::OK(); }
 
+  bool IsRemoteCallAllowed() const { return true; }
```
Comment (zhangyujing) [inline tensorflow/core/common_runtime/threadpool_device.h:47]: Is this override needed?

File: tensorflow/core/grappler/optimizers/function_optimizer.cc (hunk 4 above)
Comment (ppiskorski) [reply tensorflow/core/grappler/optimizers/function_optimizer.cc:118]: nope, will revert

File: tensorflow/core/common_runtime/threadpool_device.h (hunk 5 above)
Comment (ppiskorski) [reply tensorflow/core/common_runtime/threadpool_device.h:47]: Right, seems to be of CPU type so it's handled by whitelist.

**Comment by ppiskorski** [comment]
Is is possible to see details about CI execution? I can't reproduce the failure myself and I hope it was an issue with old code version.

File: tensorflow/core/common_runtime/device.cc (hunk 1 above)
Comment (ppiskorski) [reply tensorflow/core/common_runtime/device.cc]: To keep the change concise I eventually opted to keep base implementation of IsRemoteCallAllowed like this. Solving for VirtualDevice would snowball to extending DeviceAttributes. Keeping this allows to remove some overrides that you commented later on and provably keeps behavior of existing devices unchanged.

File: tensorflow/core/framework/device.h [hunk 6]
```diff
@@ -178,6 +178,8 @@ class Device : public DeviceBase {
 
   virtual bool IsLocal() const { return true; }
 
+  virtual bool IsRemoteCallAllowed() const;
```
Comment (zhangyujing) [inline tensorflow/core/framework/device.h:181]: Could you please add a comment for this method? Thanks!

**Review by zhangyujing — APPROVED** [review]

File: tensorflow/core/framework/device.h (hunk 6 above)
Comment (ppiskorski) [reply tensorflow/core/framework/device.h:182]: Sorry, missed that when resolving merge. Done.

Comment (ppiskorski) [reply tensorflow/core/framework/device.h:182]: Sorry, missed that when resolving merge conflict. Done.

**Review by zhangyujing — APPROVED** [review]
